### PR TITLE
image: support multi OS versions for Windows images

### LINKED ISSUE #460: Build Windows containers for multiple Windows server versions (including 2004)
**Is your feature request related to a problem?/Why is this needed**

Windows containers must match the OS version of the host/node they are running on.
Currently the container images referenced in https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/master/deploy/v0.9.0/csi-azurefile-node-windows.yaml are only built for a single Windows OS version (2019).

**Describe the solution you'd like in detail**

We should update the build process/files to build mulit-arch images with a seperate image for each Windows Server OS listed at https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#windows-os-version-support

**Describe alternatives you've considered**

**Additional context**

This PR updates the build process for the pause image to build it for multiple Windows OS versions using buildx
https://github.com/kubernetes/kubernetes/pull/91452

### PR DESCRIPTION
**Pull request by chewong**
Signed-off-by: Ernest Wong <chuwon@microsoft.com>

**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #460

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -101,52 +116,41 @@ azurefile-windows:
 azurefile-darwin:
 	CGO_ENABLED=0 GOOS=darwin go build -a -ldflags ${LDFLAGS} -o _output/azurefileplugin ./pkg/azurefileplugin
```
Comment (andyzhangx) [inline Makefile:118]: could you keep `make container` (identiclal to `make container-linux`)? thanks.
```
0.01s$ make container
make: *** No rule to make target 'container'.  Stop.
The command "make container" exited with 2.
```

File: Makefile [hunk 2]
```diff
@@ -101,52 +116,41 @@ azurefile-windows:
...
 
-.PHONY: container
-container: azurefile
-	docker build --no-cache -t $(IMAGE_TAG) -f ./pkg/azurefileplugin/dev.Dockerfile .
+.PHONY: azurefile-container-linux
```
Comment (andyzhangx) [inline Makefile:119]: remove `azurefile`, set as `container-linux`

File: Makefile [hunk 3]
```diff
@@ -101,52 +116,41 @@ azurefile-windows:
...
+azurefile-container-linux:
+	docker buildx build --pull --output=type=$(OUTPUT_TYPE) --platform="linux/$(ARCH)" \
+		-t $(IMAGE_TAG)-linux-$(ARCH) -f ./pkg/azurefileplugin/Dockerfile .
+
+.PHONY: azurefile-container-windows
```
Comment (andyzhangx) [inline Makefile:124]: `container-windows`

File: Makefile [hunk 4]
```diff
@@ -101,52 +116,41 @@ azurefile-windows:
...
+azurefile-container-windows:
+	docker buildx build --pull --output=type=$(OUTPUT_TYPE) --platform="windows/$(ARCH)" \
+		 -t $(IMAGE_TAG)-windows-$(OSVERSION)-$(ARCH) --build-arg OSVERSION=$(OSVERSION) -f ./pkg/azurefileplugin/Windows.Dockerfile .
 
-.PHONY: azurefile-container
-azurefile-container: azurefile azurefile-windows
+.PHONY: azurefile-container-all
```
Comment (andyzhangx) [inline Makefile:129]: `container-all`

File: Makefile [hunk 5]
```diff
@@ -36,6 +34,23 @@ GOBIN ?= $(GOPATH)/bin
 DOCKER_CLI_EXPERIMENTAL = enabled
 export GOPATH GOBIN GO111MODULE DOCKER_CLI_EXPERIMENTAL
 
+# Generate all combination of all OS, ARCH, and OSVERSIONS for iteration
+ALL_OS = linux windows
+ALL_ARCH.linux = amd64
+ALL_OS_ARCH.linux = $(foreach arch, ${ALL_ARCH.linux}, linux-$(arch))
+ALL_ARCH.windows = amd64
+ALL_OSVERSIONS.windows := 1809 1903 1909 2004
```
Comment (andyzhangx) [inline Makefile:42]: @AbelHu do we really need so many os versions on Windows? it's 4 versions now, and all these 4 versions are not compatible?

Comment (andyzhangx) [reply Makefile:42]: in the beginning, there is only `1809`, so what's the other `1903 1909 2004` for? Are they all necessary?

File: Makefile [hunk 6]
```diff
@@ -105,35 +122,45 @@ azurefile-darwin:
...
+			BASEIMAGE=mcr.microsoft.com/windows/nanoserver:$${osversion}; \
+			full_version=`docker manifest inspect $${BASEIMAGE} | jq -r '.manifests[0].platform["os.version"]'`; \
+			sed -i -r "s/(\"os\"\:\"windows\")/\0,\"os.version\":\"$${full_version}\"/" "${HOME}/.docker/manifests/$${manifest_image_folder}/$${manifest_image_folder}-windows-$${osversion}-$${arch}"; \
+		done; \
+	done
+	docker manifest push --purge $(IMAGE_TAG)
 	docker manifest inspect $(IMAGE_TAG)
 ifdef PUBLISH
-	docker manifest create $(IMAGE_TAG_LATEST) $(IMAGE_TAG)-linux-amd64 $(IMAGE_TAG)-windows-1809-amd64
+	docker manifest create $(IMAGE_TAG_LATEST) $(IMAGE_TAG)-linux-amd64 $(IMAGE_TAG)-windows-1809-amd64 $(IMAGE_TAG)-windows-1903-amd64 $(IMAGE_TAG)-windows-1909-amd64 $(IMAGE_TAG)-windows-2004-amd64
```
Comment (andyzhangx) [inline Makefile:161]: I did some fix on this PR, this line should be within a loop, not sure how to write such a loop in Makefile.

File: Makefile (hunk 5 above)
Comment (andyzhangx) [reply Makefile:44]: what about build `1809, 2019, 2004` as @AbelHu suggested?

Comment (chewong) [reply Makefile:44]: @marosset could you comment on this?

Comment (marosset) [reply Makefile:44]: None of the 4 versions listed here are compatible with each other. 
Let's keep all 4 of 1809, 1903, 1909, and 2004 for now so we'll include a version for each Windows Server version starting with Windows Server 2019.
This matches what gets built for the test and pause images in k/k
https://github.com/kubernetes/kubernetes/blob/ac996a37f65a8f11640808a883e9522832ed364f/build/pause/Makefile#L39

**Comment by marosset** [comment]
LGTM

File: Makefile (hunk 6 above)
Comment (chewong) [reply Makefile:161]: I think we can just do `docker tag $(IMAGE_TAG) $(IMAGE_TAG_LATEST)`

**Comment by chewong** [comment]
@andyzhangx PTAL. We would love to get some test signals on containerd + csi proxy. Thanks in advance.

**Comment by chewong** [comment]
Looks like we have already built csi-proxy v0.2.2 - https://k8sreleases.blob.core.windows.net/csi-proxy/v0.2.2/binaries/csi-proxy-v0.2.2.tar.gz

File: Makefile (hunk 6 above)
Comment (andyzhangx) [reply Makefile]: I used `docker manifest create $(IMAGE_TAG_LATEST) $(foreach osarch, $(ALL_OS_ARCH), $(IMAGE_TAG)-${osarch})`, other part LGTM

Comment (chewong) [reply Makefile]: LGTM

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve

**Comment by andyzhangx** [comment]
@chewong  could you also make same change on azure disk driver? thanks.

**Pull request merged**